File: BackEnd/src/domain/interfaces/usecases/chat/ISendMessageUseCase.ts
/**
 * @file ISendMessageUseCase.ts
 * @description Interfaz del caso de uso para enviar un mensaje en una conversación.
 * @module Chat
 * @layer Domain
 */

import type { SendMessageRequestDto } from '../../../dtos/chat/send-message-request.dto.js';
import type { MessageResponseDTO } from '../../../dtos/chat/message-response.dto.js';

/**
 * Contrato del use case SendMessage.
 * Persiste el mensaje y emite el evento Socket.IO al destinatario.
 */
export interface ISendMessageUseCase {
  /**
   * Ejecuta el caso de uso.
   *
   * @param conversationId — ID de la conversación donde se envía el mensaje.
   * @param senderId — ID del usuario emisor.
   * @param dto — Datos del mensaje (texto, contentMeta, tempId).
   * @returns MessageResponseDTO del mensaje persistido.
   * @throws {NotFoundException} Si la conversación no existe.
   * @throws {ForbiddenException} Si el emisor no es participante.
   */
  execute(conversationId: string, senderId: string, dto: SendMessageRequestDto): Promise<MessageResponseDTO>;
}

File: FrontEnd/lib/domain/interfaces/usecases/chat/i_send_message_use_case.dart
/// @file i_send_message_use_case.dart
/// @description Interfaz: Envía un mensaje en una conversación con soporte optimista.
/// @module Chat
/// @layer Domain
library;

import '../../../entities/message.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// I SEND MESSAGE USE CASE
// ═══════════════════════════════════════════════════════════════════════════════

/// Contrato del use case que envía un mensaje en una conversación.
///
/// Soporta actualizaciones optimistas mediante [tempId]: el cliente asigna
/// un ID temporal antes de enviar y lo usa para reconciliar el mensaje
/// cuando llega la confirmación del servidor.
abstract interface class ISendMessageUseCase {
  /// Envía [content] en [conversationId].
  ///
  /// [param] conversationId — ID de la conversación destino.
  /// [param] content — texto del mensaje.
  /// [param] tempId — ID temporal del cliente para reconciliación optimista (opcional).
  /// [returns] [Message] confirmado (o con [MessageStatus.pending] si offline).
  /// [throws] AppError — si el envío falla por error distinto a red.
  Future<Message> execute(
    String conversationId,
    String content, {
    String? tempId,
  });
}
